### DISCUSSION TITLE
Determine running state of Codespaces and stop from /codespaces UI

### DISCUSSION THREAD

**Discussion by jamesez** [community/community#5203] [Codespaces]
I guess it's because I went to Montessori or something but I `docker compose stop` and close all my windows and tabs at the end of the day.

There doesn't seem? to be a reported state of whether a Codespace is running in the Codespaces list. I know we're in free preview, but I would like to know if the taxi meter is still running.  I might have e.g. foolishly left it plugged in on an webhook config and it's getting events but nobody is there.

Could you consider adding an icon (green dot, plugged in, say) to indicate that a Codespaces is up, and perhaps a "stop codespace" item to the actions menu?

Thank you!

**Comment by tanmayeekamath** [comment]
Hey @jamesez, thanks for your feedback. 

I wanted to share a few resources below, just in case :) 
You can view Codespaces (and related information) created across different repos by going to https://github.com/codespaces. 
Codespaces are manually stoppable using the command palette in VS Code: https://docs.github.com/en/codespaces/codespaces-reference/using-the-command-palette-in-codespaces#suspending-or-stopping-a-codespace
Codespaces will also automatically stop/suspend after 30 mins of idle activity to minimize wasteful resources. 

To help me capture the request, can you confirm if my understanding is correct? 
You would like to view the state of the Codespace (running, suspended) [in this view](https://github.com/codespaces) and also be able to stop the Codespace from that page. 

Thanks!

**Reply by jamesez** [reply]
Yep, you have it right - please show running / suspended on https://github.com/codespaces and additionally provide a way to stop them. 

While I can stop it in the command palette, that requires launching the space - perhaps starting it by accident - just to then stop it.

As to automatic suspension, I found that the docs are quite vague. One part says `terminal output are counted as activity and so your codespace will not stop if terminal output is continuing` which is a bit confusing: if I left a process running that emits a status heartbeat (say it looks for work to do every so often), will the codespace suspend if I close all editors? How quickly after I disconnect all editors? What if I left VSCode up on my desktop and locked the screen and forgot?

**Reply by tanmayeekamath** [reply]
- If I left a process running that emits a status heartbeat (say it looks for work to do every so often), will the codespace suspend if I close all editors?
**Answer**: Can you define what 'work' would entail? If it outputs something to the terminal or writes to a file, that should count as activity and hence the Codespace will not auto-suspend. 
- How quickly after I disconnect all editors?
**Answer**: 30 minutes after you disconnect the editor assuming there aren't other activities happening. - e.g. output being written to the terminal by a process.
- What if I left VSCode up on my desktop and locked the screen and forgot? 
**Answer**: Codespace will auto suspend after 30 mins in this case.

**Reply by jablonski** [reply]
>       **Answer**: Can you define what 'work' would entail? If it outputs something to the terminal or writes to a file, that should count as activity and hence the Codespace will not auto-suspend.

This might cause trouble, because developers are lazy people and don't think about costs. Think of a simple task which logs some details every minute... this Codespace would run forever.

What is the reasoning behind "write to file & log to terminal" not to auto-suspend a Codespace?

At least a global setting for an organization would be cool if you don't like the default behaviour:  "Auto-suspend Codespace after 30 minutes if no VSCode is attached."

**Reply by jablonski** [reply]
Other ideas for preventing unbounded costs for "forgotten" Codespaces would be:

- Allow an organization to stop all Codespaces which are running more than x hours (e.g. 12 hours, no developer should work longer per day)
- Allow an organization to schedule a "Stop all Codespaces"-Task... e.g. 3am every day.
- Allow an organization to define a budget per developer, so every developer considers running Codespaces as precious resources.

To use Codespaces as an organization we need more control of the Codespace usage in an organization. At the moment the only possible way is to define an overall budget... we need more fine grained control.
